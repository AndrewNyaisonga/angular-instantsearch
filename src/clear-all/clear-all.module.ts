import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgISFooterModule } from "../footer/footer.module";
import { NgISHeaderModule } from "../header/header.module";
import { NgISClearAll } from "./clear-all";

@NgModule({
  declarations: [NgISClearAll],
  entryComponents: [NgISClearAll],
  exports: [NgISClearAll],
  imports: [CommonModule, NgISHeaderModule, NgISFooterModule]
})
export class NgISClearAllModule {}
