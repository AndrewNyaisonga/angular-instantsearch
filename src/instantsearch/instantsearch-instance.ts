import { Injectable } from "@angular/core";
import instantsearch from "instantsearch.js/es";

@Injectable()
export class NgISInstance {
  public appId: string;
  public apiKey: string;
  public indexName: string;

  public numberLocale?: string;
  public searchFunction?: () => void;
  public createAlgoliaClient?: () => object;
  public searchParameters?: object | void;
  public urlSync?:
    | boolean
    | {
        mapping?: object;
        threshold?: number;
        trackedParameters?: string[];
        useHash?: boolean;
        getHistoryState?: () => object;
      };

  private instance?: {
    start: () => void;
  };

  public init(config: object) {
    this.instance = instantsearch(config);
  }

  public start() {
    this.instance.start();
  }
}
