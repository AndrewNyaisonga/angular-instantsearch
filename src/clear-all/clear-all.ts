import { Component, Input } from "@angular/core";
import { connectClearAll } from "instantsearch.js/es/connectors";
import { noop } from "lodash";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";
import { bem } from "../utils";

const cx = bem("ClearAll");

@Component({
  selector: "ng-ais-clear-all",
  template: `
    <div class="${cx()}">
      <ng-ais-header [header]="header" className="${cx(
        "header"
      )}"></ng-ais-header>

      <div class="${cx("body")}">
        <button
          class="${cx("button")}"
          (click)="handleClick($event)"
          disable="!state.hasRefinements"
        >
          {{buttonLabel}}
        </button>
      </div>

      <ng-ais-footer [footer]="footer" className="${cx(
        "footer"
      )}"></ng-ais-footer>
    </div>
  `
})
export class NgAisClearAll extends BaseWidget {
  @Input() public buttonLabel: string = "Clear all";
  @Input() public clearsQuery: boolean = false;
  @Input() public excludeAttributes: string[] = [];

  public state = { hasRefinements: false, refine: noop };

  constructor(searchInstance: NgAisInstance) {
    super(searchInstance);
  }

  public ngOnInit() {
    // we need to `createWidget` from `ngOnInit` to have `@Input()` intialized
    this.createWidget(connectClearAll, {
      clearsQuery: this.clearsQuery,
      excludeAttributes: this.excludeAttributes
    });

    super.ngOnInit();
  }

  public handleClick(event) {
    event.preventDefault();

    if (this.state.hasRefinements) {
      this.state.refine();
    }
  }
}
