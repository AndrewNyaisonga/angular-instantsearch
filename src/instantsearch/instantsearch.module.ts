import { CommonModule } from "@angular/common";
import { ModuleWithProviders, NgModule } from "@angular/core";

import { NgISInstantSearch } from "./instantsearch";
import { NgISInstance } from "./instantsearch-instance";

@NgModule({
  declarations: [NgISInstantSearch],
  entryComponents: [NgISInstantSearch],
  exports: [NgISInstantSearch],
  imports: [CommonModule]
})
export class NgISInstantSearchModule {
  public static forRoot(): ModuleWithProviders {
    return {
      ngModule: NgISInstantSearchModule,
      providers: [NgISInstance]
    };
  }
}
