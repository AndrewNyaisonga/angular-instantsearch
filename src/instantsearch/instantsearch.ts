import { AfterViewInit, Component, Input, OnInit } from "@angular/core";

import { NgISInstance } from "./instantsearch-instance";

@Component({
  selector: "ngis-instantsearch",
  template: `<ng-content></ng-content>`
})
export class NgISInstantSearch implements AfterViewInit, OnInit {
  @Input() public config: InstantSearchConfig;

  constructor(private searchInstance: NgISInstance) {}

  public ngOnInit() {
    this.searchInstance.init(this.config);
  }

  public ngAfterViewInit() {
    this.searchInstance.start();
  }
}
