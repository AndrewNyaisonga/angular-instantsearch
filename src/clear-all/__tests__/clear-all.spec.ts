import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../../instantsearch/instantsearch.module";
import { NgAisClearAllModule } from "../clear-all.module";
import { NgAisClearAll } from "../clear-all";

jest.mock("../../base-widget");

describe("ClearAll", () => {
  beforeEach(() =>
    TestBed.configureTestingModule({
      declarations: [],
      imports: [NgAisInstantSearchModule.forRoot(), NgAisClearAllModule]
    })
  );

  it("renders markup without state", () => {
    const fixture = TestBed.createComponent(NgAisClearAll);
    expect(fixture).toMatchSnapshot();
  });

  it("should render disabled button when no refinements", () => {
    const fixture = TestBed.createComponent(NgAisClearAll);
    fixture.detectChanges();

    const btn = fixture.debugElement.nativeElement.querySelector("button");
    expect(btn.disabled).toBeTruthy();
  });

  it("should render enabled button when refinements", () => {
    const fixture = TestBed.createComponent(NgAisClearAll);

    fixture.componentInstance.updateState(
      {
        hasRefinements: true,
        refine: jest.fn()
      },
      false
    );
    fixture.detectChanges();

    const btn = fixture.debugElement.nativeElement.querySelector("button");
    expect(btn.disabled).toBeFalsy();
  });

  it("should call refine() when clicked on button", () => {
    const spy = jest.fn();
    const fixture = TestBed.createComponent(NgAisClearAll);

    fixture.componentInstance.updateState(
      {
        hasRefinements: true,
        refine: spy
      },
      false
    );
    fixture.detectChanges();

    const btn = fixture.debugElement.nativeElement.querySelector("button");
    btn.click();

    expect(spy).toHaveBeenCalled();
  });
});
