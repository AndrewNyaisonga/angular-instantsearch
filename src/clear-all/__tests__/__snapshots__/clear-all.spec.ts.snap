// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearAll renders markup without state 1`] = `
<ng-ais-clear-all
  buttonLabel={[Function String]}
  clearsQuery="false"
  excludeAttributes={[Function Array]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-ClearAll"
  >
    
        
    <ng-ais-header
      classname="ais-ClearAll-header ais-header"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-ClearAll-body ais-body"
    >
      
          
      <button
        class="ais-ClearAll-button"
      >
        
            
      </button>
      
        
    </div>
    
  
        
    <ng-ais-footer
      classname="ais-ClearAll-footer ais-footer"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-clear-all>
`;
